File: python/examples/backend/providers/vertexai.py
import asyncio

from pydantic import BaseModel, Field

from beeai_framework.adapters.vertexai.backend.chat import VertexAIChatModel
from beeai_framework.backend.chat import ChatModel
from beeai_framework.backend.events import ChatModelNewTokenEvent
from beeai_framework.backend.message import UserMessage
from beeai_framework.cancellation import AbortSignal
from beeai_framework.emitter import EventMeta
from beeai_framework.errors import AbortError
from beeai_framework.parsers.field import ParserField
from beeai_framework.parsers.line_prefix import LinePrefixParser, LinePrefixParserNode


async def vertexai_from_name() -> None:
    llm = ChatModel.from_name("vertexai:gemini-2.0-flash-lite-001")
    user_message = UserMessage("what states are part of New England?")
    response = await llm.create(messages=[user_message])
    print(response.get_text_content())


async def vertexai_sync() -> None:
    llm = VertexAIChatModel("gemini-2.0-flash-lite-001")
    user_message = UserMessage("what is the capital of Massachusetts?")
    response = await llm.create(messages=[user_message])
    print(response.get_text_content())


async def vertexai_stream() -> None:
    llm = VertexAIChatModel("gemini-2.0-flash-lite-001")
    user_message = UserMessage("How many islands make up the country of Cape Verde?")
    response = await llm.create(messages=[user_message], stream=True)
    print(response.get_text_content())


async def vertexai_stream_abort() -> None:
    llm = VertexAIChatModel("gemini-2.0-flash-lite-001")
    user_message = UserMessage("What is the smallest of the Cape Verde islands?")

    try:
        response = await llm.create(messages=[user_message], stream=True, abort_signal=AbortSignal.timeout(0.5))

        if response is not None:
            print(response.get_text_content())
        else:
            print("No response returned.")
    except AbortError as err:
        print(f"Aborted: {err}")


async def vertexai_structure() -> None:
    class TestSchema(BaseModel):
        answer: str = Field(description="your final answer")

    llm = VertexAIChatModel("gemini-2.0-flash-lite-001")
    user_message = UserMessage("How many islands make up the country of Cape Verde?")
    response = await llm.create_structure(schema=TestSchema, messages=[user_message])
    print(response.object)


async def vertexai_stream_parser() -> None:
    llm = VertexAIChatModel("gemini-2.0-flash-lite-001")

    parser = LinePrefixParser(
        nodes={
            "test": LinePrefixParserNode(
                prefix="Prefix: ", field=ParserField.from_type(str), is_start=True, is_end=True
            )
        }
    )

    async def on_new_token(data: ChatModelNewTokenEvent, event: EventMeta) -> None:
        await parser.add(data.value.get_text_content())

    user_message = UserMessage("Produce 3 lines each starting with 'Prefix: ' followed by a sentence and a new line.")
    await llm.create(messages=[user_message], stream=True).observe(
        lambda emitter: emitter.on("new_token", on_new_token)
    )
    result = await parser.end()
    print(result)


async def main() -> None:
    print("*" * 10, "vertexai_from_name")
    await vertexai_from_name()
    print("*" * 10, "vertexai_sync")
    await vertexai_sync()
    print("*" * 10, "vertexai_stream")
    await vertexai_stream()
    print("*" * 10, "vertexai_stream_abort")
    await vertexai_stream_abort()
    print("*" * 10, "vertexai_structure")
    await vertexai_structure()
    print("*" * 10, "vertexai_stream_parser")
    await vertexai_stream_parser()


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/examples/backend/providers/vertexai.ts
import "dotenv/config.js";
import { GoogleVertexChatModel } from "beeai-framework/adapters/google-vertex/backend/chat";
import { ToolMessage, UserMessage } from "beeai-framework/backend/message";
import { ChatModel } from "beeai-framework/backend/chat";
import { z } from "zod";
import { ChatModelError } from "beeai-framework/backend/errors";
import { OpenMeteoTool } from "beeai-framework/tools/weather/openMeteo";

const llm = new GoogleVertexChatModel(
  "gemini-1.5-flash-001",
  // {},
  // {
  //   location: "GOOGLE_VERTEX_LOCATION",
  //   project: "GOOGLE_VERTEX_PROJECT",
  //   googleAuthOptions: {
  //     apiKey: "",
  //   },
  // },
);

llm.config({
  parameters: {
    topK: 1,
    temperature: 0,
    topP: 1,
  },
});

async function googleVertexFromName() {
  const googleVertexLLM = await ChatModel.fromName("google-vertex:gemini-1.5-flash-001");
  const response = await googleVertexLLM.create({
    messages: [new UserMessage("what states are part of New England?")],
  });
  console.info(response.getTextContent());
}

async function googleVertexSync() {
  const response = await llm.create({
    messages: [new UserMessage("what is the capital of Massachusetts?")],
  });
  console.info(response.getTextContent());
}

async function googleVertexStream() {
  const response = await llm.create({
    messages: [new UserMessage("How many islands make up the country of Cape Verde?")],
    stream: true,
  });
  console.info(response.getTextContent());
}

async function googleVertexAbort() {
  try {
    const response = await llm.create({
      messages: [new UserMessage("What is the smallest of the Cape Verde islands?")],
      stream: true,
      abortSignal: AbortSignal.timeout(1 * 1000),
    });
    console.info(response.getTextContent());
  } catch (err) {
    if (err instanceof ChatModelError) {
      console.error("Aborted", { err });
    }
  }
}

async function googleVertexStructure() {
  const response = await llm.createStructure({
    schema: z.object({
      answer: z.string({ description: "your final answer" }),
    }),
    messages: [new UserMessage("How many islands make up the country of Cape Verde?")],
  });
  console.info(response.object);
}

async function googleVertexToolCalling() {
  const userMessage = new UserMessage("What is the weather in Boston?");
  const weatherTool = new OpenMeteoTool({ retryOptions: { maxRetries: 3 } });
  const response = await llm.create({ messages: [userMessage], tools: [weatherTool] });
  const toolCallMsg = response.getToolCalls()[0];
  console.debug(JSON.stringify(toolCallMsg));
  const toolResponse = await weatherTool.run(toolCallMsg.args as any);
  const toolResponseMsg = new ToolMessage({
    type: "tool-result",
    result: toolResponse.getTextContent(),
    toolName: toolCallMsg.toolName,
    toolCallId: toolCallMsg.toolCallId,
  });
  console.info(toolResponseMsg.toPlain());
  const finalResponse = await llm.create({ messages: [userMessage, toolResponseMsg], tools: [] });
  console.info(finalResponse.getTextContent());
}

async function googleVertexDebug() {
  // Log every request
  llm.emitter.match("*", (value, event) =>
    console.debug(
      `Time: ${event.createdAt.toISOString()}`,
      `Event: ${event.name}`,
      `Data: ${JSON.stringify(value)}`,
    ),
  );

  const response = await llm.create({
    messages: [new UserMessage("Hello world!")],
  });
  console.info(response.messages[0].toPlain());
}

console.info(" googleVertexFromName".padStart(25, "*"));
await googleVertexFromName();
console.info(" googleVertexSync".padStart(25, "*"));
await googleVertexSync();
console.info(" googleVertexStream".padStart(25, "*"));
await googleVertexStream();
console.info(" googleVertexAbort".padStart(25, "*"));
await googleVertexAbort();
console.info(" googleVertexStructure".padStart(25, "*"));
await googleVertexStructure();
console.info(" googleVertexToolCalling".padStart(25, "*"));
await googleVertexToolCalling();
console.info(" googleVertexDebug".padStart(25, "*"));
await googleVertexDebug();
